internal/precache/workload/pullImages.go

```diff
@@ -0,0 +1,200 @@
+		return false
+	}
+	return true
+}
+
+// podmanImgExists reports the existence of the given image via podman CLI
+func podmanImgExists(image string) bool {
+	if err := executeCmd(true, "podman", []string{"image", "exists", image}...); err != nil {
+		return false
+	}
+	return true
+}
+
+// podmanImgPull pulls the specified image via podman CLI
+func podmanImgPull(image, authFile string) error {
+	args := []string{"pull", image}
+	if authFile != "" {
+		args = append(args, []string{"--authfile", authFile}...)
+	}
+	return executeCmd(false, "podman", args...)
+}
+
+// pullImage attempts to pull an image via podman CLI
+func pullImage(image, authFile string, progress *precache.Progress) error {
+
+	var err error
+	for i := 0; i < MaxRetries; i++ {
+		err = podmanImgPull(image, authFile)
+		if err == nil {
+			log.Infof("Successfully pulled image: %s", image)
+			break
+		} else {
+			log.Infof("Attempt %d/%d: Failed to pull %s: %v", i+1, MaxRetries, image, err)
+		}
+	}
+	// update precache progress tracker
+	progress.Update(err == nil, image)
+
+	// persist progress to file
+	progress.Persist(precache.StatusFile)
+
+	return err
+}
+
+// getAuthFile returns the auth file for podman
+func getAuthFile() (string, error) {
+	// Configure Podman auth file
+	authFile := os.Getenv(EnvAuthFile)
+	if authFile == "" {
+		authFile = DefaultAuthFile
+	}
+
+	// Check if authFile exists
+	if _, err := os.Stat(authFile); os.IsNotExist(err) {
+		log.Errorf("Missing auth file for podman")
+		return "", err
+	}
+	log.Info("Auth file for podman found.")
+
+	return authFile, nil
+}
+
+// PullImages pulls a list of images using podman
+func PullImages(precacheSpec []string) (progress *precache.Progress, err error) {
+
+	// Get auth file for Podman
+	authFile, err := getAuthFile()
+	if err != nil {
+		return progress, err
+	}
+
+	// Initialize progress tracking
+	progress = &precache.Progress{
+		Total:   len(precacheSpec),
+		Pulled:  0,
+		Skipped: 0,
+		Failed:  0,
+	}
+
+	var pullSpec = make([]string, 0, len(precacheSpec))
+	// Sift through image list to determine which images exist, and which need to be pulled
+	log.Infof("Checking the pre-cache spec file images to determine if they need to be pulled...")
+	var skip bool
+	for _, image := range precacheSpec {
+		skip = podmanImgExists(image)
+		if !skip {
+			pullSpec = append(pullSpec, image)
+		} else {
+			log.Infof("%s exists, skipping it...", image)
+			progress.Skipped++
+		}
+	}
+	log.Infof("Check complete: %d images need to be pulled!", len(pullSpec))
+
+	// Create wait group and pull images
+	var wg sync.WaitGroup
+	numThreads, err := strconv.Atoi(os.Getenv(precache.EnvMaxPullThreads))
+	if err != nil {
+		numThreads = precache.DefaultMaxConcurrentPulls
+	}
+	threads := make(chan struct{}, numThreads)
+	log.Infof("Configured precaching job to concurrently pull %d images.", numThreads)
+
+	// Start pulling images
+	for _, image := range pullSpec {
```

Comment:
nit: maybe pullSpecs (plural)?

Reply:
Spec refers to a specification object, in this context, the specification object is a list of images. The relationship is 1 pullSpec : 1 list of images. So plural would imply multiple lists of images.